mansrc/load2sqlitedb.1
'\" t
.\"     Title: load2sqlitedb
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "LOAD2SQLITEDB" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
load2sqlitedb \- load genome sequences and extrinsic evidence hints into a SQLite database
.SH "SYNOPSIS"
.sp
\fBload2sqlitedb\fP [OPTIONS] \-\-species=\fISPECIES\fP \-\-dbaccess=\fIdatabase.db\fP \fIinputfilename\fP
.SH "DESCRIPTION"
.sp
\fBload2sqlitedb\fP is a tool to load genome sequences and extrinsic evidence hints into a SQLite database.
.br
When storing genomes/hints of multiple organisms call this program repeatedly for each one.
.SH "OPTIONS"
.sp
\fIinputfilename\fP refers to a genome file in FASTA format or a hints file in GFF format.
.SS "Mandatory options"
.sp
\fB\-s\fP, \fB\-\-species\fP=\fISPECIES\fP
.RS 4
SPECIES is a name for your species and the same identifier as is used in the treefile and alnfile parameters to augustus.
.RE
.sp
\fB\-d\fP, \fB\-\-dbaccess\fP=\fIdatabase.db\fP
.RS 4
The file name of the SQLite database that will be opened or created if it does not already exist.
.RE
.SS "Additional options"
.sp
\fB\-c\fP, \fB\-\-chunksize\fP=\fIsize\fP
.RS 4
This option is only relevant when loading a sequence file.
.br
The sequences in the input genome are split into chunks of this size so
that subsequent retrievals of small sequence ranges do not require to read
the complete \- potentially much longer \- chromosome. (\(lA 1000000). Default Value: 50000
.RE
.sp
\fB\-i\fP, \fB\-\-noIdx\fP
.RS 4
Use this flag to suppress the building of indices on the database tables.
.br
If you are going to load several genomes and/or hint files in a row, this option
is recommended to speed up the loading. But make sure to build indices with
\-\-makeIdx after all genomes/hints are loaded. Otherwise, data retrieval operations
can be very slow.
.RE
.sp
\fB\-m\fP, \fB\-\-makeIdx\fP
.RS 4
Use this flag to build the indices on the database tables after loading several genomes and/or hint files with \-\-noIdx.
.br
Only call this once for all species, e.g. load2sqlitedb \-\-makeIdx \-\-dbaccess=database.db
.RE
.sp
\fB\-r\fP, \fB\-\-clean\fP
.RS 4
Makes a clean load deleting existing hints/genome for the species from the database.+
When called with a gff file, only the hints for the species are delete, but not the genome.
.br
When called with a fasta file, both hints and genome for the species are deleted.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Produce help message.
.RE
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
  load2sqlitedb \-\-species=mouse \-\-dbaccess=vertebrates.db mouse.fa
  load2sqlitedb \-\-species=mouse \-\-dbaccess=vertebrates.db mouse.hints.gff
  load2sqlitedb \-\-species=human \-\-dbaccess=vertebrates.db human.fa
  load2sqlitedb \-\-species=human \-\-dbaccess=vertebrates.db human.hints.gff
.fi
.if n \{\
.RE
.\}
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

src/load2sqlitedb.cc
/*
 * load2sqlitedb.cc
 *
 * License: Artistic License, see file LICENSE.TXT or 
 *          https://opensource.org/licenses/artistic-license-1.0
 * 
 * Description: Bulk loading of sequences and hints into an SQLite database.
 */

// Project includes
#include "fasta.hh"
#include "hints.hh"
#include "sqliteDB.hh"
#include "projectio.hh"

// standard C/C++ includes
#include <string>
#include <iostream>
#include <fstream>
#include <getopt.h>     /* for getopt_long; standard getopt is in unistd.h */
#include <stdlib.h>     /* for exit() */
#include <exception>

#include <stdio.h>

using namespace std;

int chunksize = 50000;

void printUsage();

/*
 * main
 */
int main( int argc, char* argv[] ){
    int c;
    int help = 0;
    int noIdx = 0;
    int makeIdx = 0;
    bool clean = false;
    string species;
    string dbfile;
    string fastafile;

    static struct option long_options[] = {
        {"species", 1, 0, 's'},
        {"dbaccess", 1, 0, 'd'},
        {"help", 0, 0, 'h'},
	{"chunksize", 1, 0, 'c'},
        {"noIdx", 0, 0, 'i'},
        {"makeIdx", 0, 0, 'm'},
	{"clean", 0, 0, 'r'},
        {NULL, 0, NULL, 0}
    };
    int option_index = 0;
    while ((c = getopt_long(argc, argv, "s:d:hc:imr", long_options, &option_index)) != -1) {
        switch (c) {
        case 's':
	    species = optarg;
            break;
        case 'd':
	    dbfile = optarg;
            break;
        case 'h':
            help = 1;
            break;
        case 'c':
	    chunksize = atoi(optarg);
            break;
	case 'i':
            noIdx = 1;
	    break;
	case 'm':
            makeIdx = 1;
	    break;
	case 'r':
	    clean = true;
        default:
            break;
        }
    }
    if (dbfile.empty()){
	cerr << "Missing database file. dbaccess is a required parameter." << endl;
	printUsage();
	exit(1);
    }
    if (noIdx && makeIdx){
	cerr << "You can only use one of the options --noIdx or --makeIdx at a time." << endl;
	exit(1);
    }
    if(makeIdx){ // only build indices
	SQLiteDB db(dbfile.c_str(),crw);
	try{
	    cout << "Creating indices on genomes tables."<< endl;	    
	    db.exec("CREATE INDEX IF NOT EXISTS gidx ON genomes(speciesid,seqnr,start,end);");
	    if(db.tableExists("hints")){
		cout << "Creating indices on hints tables."<< endl;	    		
		db.exec("CREATE INDEX IF NOT EXISTS hidx ON hints(speciesid,seqnr,start,end);");
	    }
	}catch(const char* err){
	    cerr << "Failed creating indices." << endl;
	    cerr << err << endl;
	    exit(1);
	}
	exit(0);
    }
    if (optind < argc-2) {
        cerr << "More than two options without name: ";
        while (optind < argc)
            cerr << " " << argv[optind++];
	cerr << endl << "Only the sequence/hints file name does not require a parameter name." << endl;
	if (!help)
	    printUsage();
	exit(1);
    }
    if (optind == argc-1){
	fastafile = argv[optind];
	// expand the ~ to the $Home directory
	if (fastafile.length()>0 && fastafile[0]=='~')
	    fastafile.replace(0,1,getenv("HOME"));
    }
    else {
	cerr << "Missing sequence/hints file name." << endl;
	printUsage();
	exit(1);
    }
    if (help) {
	printUsage();
	exit(1);
    }
    if (species.empty()){
	cerr << "Missing species name. Required parameter." << endl;
	printUsage();
	exit(1);
    }
    if (chunksize < 2){
	cerr << "Chunksize too small (" << chunksize << "). Should be roughly in the order of a gene's length." << endl;
	printUsage();
	exit(1);
    }
    if (chunksize > 1000000){
	cerr << "Chunksize too big (" << chunksize << "). " << endl;
	printUsage();
	exit(1);
    }

    // open database
    SQLiteDB db(dbfile.c_str(),crw);
    try {
	db.createTableSpeciesnames();
	db.createTableSeqnames();
	db.createTableGenomes();
	db.createTableHints();
	db.createTableFeatureTypes();

	ifstream ifstrm;
	ifstrm.open(fastafile.c_str());
	if( !ifstrm ){
	    cerr << "Could not open input file " << fastafile << endl;
	    exit(1);
	}
	// if input file is in Fasta format, try to load sequences
	if(isFasta(ifstrm)){
	    cout << "Looks like " << fastafile << " is in fasta format." << endl;

	    // drop index on genomes table for faster insertion
	    db.exec("DROP INDEX IF EXISTS gidx;");

	    int seqCount = 0, chunkCount = 0;
	    unsigned int lenCount = 0;
	    int speciesid = db.getSpeciesID(species,clean);
	
	    db.beginTransaction();

	    Statement stmt1(&db); 
	    stmt1.prepare("INSERT INTO genomes VALUES(?1,?2,?3,?4,?5,?6,?7);"); // prepared statement for inserting sequence chunks

	    Statement stmt2(&db);
	    stmt2.prepare("INSERT INTO seqnames (speciesid,seqname) VALUES(?1,?2);"); // prepared statement for inserting fasta headers
	    	    
	    while (ifstrm){
		char *name = NULL;
		readFastaHeader(ifstrm, name);
		try{
		    stmt2.bindInt(1,speciesid);
		    stmt2.bindText(2,name);
		    stmt2.step();
		    stmt2.reset();
		}catch(const char* err){
		    cerr << "failed inserting sequence "<< name << " for species "<< species <<" (ID="<<speciesid<<")"<< endl;
		    cerr << "Is it possible that the sequence is already in the database?" << endl;
		    cerr << "If you want to replace the existing genome, please make a clean load with option --clean" << endl;
		    throw err;
		}
		int seqnr = db.lastInsertID();
		int start = 0;
		int length = 0;
		string line;
		char c;
		
		streampos file_start = ifstrm.tellg();
		streampos file_end = file_start;
		streampos pos = file_end;
	
		while (ifstrm && ifstrm.peek() != '>' && ifstrm.peek() != EOF){
		    ifstrm.get(c);
		    pos += streamoff(1);
		    if(isalpha(c)){
			length++;
			file_end = pos;
		    }
		    if(length >= chunksize){
			stmt1.bindInt(2,seqnr);
			stmt1.bindInt(3,speciesid);
			stmt1.bindInt(4,start);
			stmt1.bindInt(5,start+length-1);
			stmt1.bindInt64(6,(uint64_t)file_start);
			stmt1.bindInt(7,file_end-file_start);
			stmt1.step();
			stmt1.reset();
  
			chunkCount++;
			lenCount +=length;

			start += length;
			// next file_start position is the character preceeding
			// the next non-whitespace character
			while (ifstrm && !isalpha(ifstrm.peek()) && ifstrm.peek() != '>' && ifstrm.peek() != EOF){
			    ifstrm.get(c);
			    pos += streamoff(1);
			}
			file_start = pos;
			length = 0;
		    }
		}			
		// last chunk
		if(length > 0){
		    stmt1.bindInt(2,seqnr);
		    stmt1.bindInt(3,speciesid);
		    stmt1.bindInt(4,start);
		    stmt1.bindInt(5,start+length-1);
		    stmt1.bindInt64(6,(uint64_t)file_start);
		    stmt1.bindInt(7,file_end-file_start);
		    stmt1.step();
		    stmt1.reset();
		
		    chunkCount++;
		    lenCount +=length;
		}
		delete name;
		seqCount++;
		
	    }
	    db.endTransaction();
		
	    // rebuild index on genomes table
	    if(!noIdx)
		db.exec("CREATE INDEX gidx ON genomes(speciesid,seqnr,start,end);");

	    if (seqCount > 0)
		cout << "Inserted " << chunkCount << " chunks of " << seqCount << " sequences (total length "
		     << lenCount << " bp)." << endl;
	    else
		cout << "No sequences found. Nothing inserted into database." << endl;
	        
	}
	else if(isGFF(ifstrm)){ // if input file is in GFF format, try to load hints
	    cout << "Looks like " << fastafile << " is in gff format." << endl;

	    int hintCount=0; // number of hints inserted into the database
	    
	    // drop index on hints table (for faster insertion)
	    db.exec("DROP INDEX IF EXISTS hidx");

	    if(clean){
	      int id = db.getSpeciesID(species,false,true);
	      if(id >= 0){
		db.deleteHints(id); // delete existing hints from DB
	        cout << "Deleted existing hints for " << species << " from database." << endl;
	      }
	    }
	    // bulk insert of all hints
	    db.beginTransaction();

	    Statement stmt(&db);
	    stmt.prepare("INSERT INTO hints (speciesid,seqnr,source,start,end,score,type,strand,frame,priority,grp,mult,esource) \
		       select speciesid,seqnr,?3,?4,?5,?6,?7,?8,?9,?10,?11 \
                       ,?12,?13 from speciesnames natural join seqnames \
                       where speciesname=?1 and seqname=?2;");
	    
	    while(ifstrm){
		Feature f;
		ifstrm >> f >> comment >> ws;
		try{
		    stmt.bindText(1,species.c_str());
		    stmt.bindText(2,f.seqname.c_str());
		    stmt.bindText(3,f.source.c_str());
		    stmt.bindInt(4,f.start);
		    stmt.bindInt(5,f.end);
		    stmt.bindDouble(6,f.score);
		    stmt.bindInt(7,f.type);
		    switch (f.strand) {
		    case plusstrand : stmt.bindText(8,"+"); break;
		    case minusstrand : stmt.bindText(8,"-"); break;
		    default : stmt.bindText(8,".");
		    }
		    switch (f.frame) {
		    case 0 :  stmt.bindText(9,"0"); break;
		    case 1 :  stmt.bindText(9,"1"); break;
		    case 2 :  stmt.bindText(9,"2"); break;
		    default :  stmt.bindText(9,".");
		    }
		    stmt.bindInt(10,f.priority);
		    stmt.bindText(11,f.groupname.c_str());
		    stmt.bindInt(12,f.mult);
		    stmt.bindText(13,f.esource.c_str());
		    stmt.step();
		    stmt.reset();
		}catch(const char* error){
		    cerr << "insert failed on\n" << f << endl;
		    cerr << error << endl;
		    exit(1);
		}
		if(db.numChanges() != 1){
		    cerr << "insert failed on\n" << f << endl;
		    cerr << "hints can only be inserted for sequences in the database." << endl;
			exit(1);
		}
		hintCount++;
	    }	
	    db.endTransaction();
	    
	    // rebuild index on hints table
	    if(!noIdx)
		db.exec("CREATE INDEX hidx ON hints(speciesid,seqnr,start,end);");

	    if(hintCount > 0)
		cout << "Inserted " << hintCount<< " hints for species " << species << endl; 
	    else
		cout << "No hints found. Nothing inserted into database." << endl;
	}
	else{
	    cout << fastafile << " is neither in gff nor in fasta format." << endl;
	}
	ifstrm.close();   
    } catch(const char* err){
        cerr << "\n" <<  argv[0] << ": ERROR\n\t" << err << "\n\n";
	exit(1);
    }
    exit(0);
}

void printUsage(){
    cout << "usage:\n\
load2sqlitedb [parameters] --species=SPECIES --dbaccess=database.db fastafile\n\
\n\
fastafile refers to a genome file in FASTA format or a hints file in GFF format\n\
SPECIES is the same identifier as is used in the treefile and alnfile parameters to augustus.\n\
\n\
database.db is the name of the database that will be opened or created if it does not exist already.\n\
When storing genomes/hints of multiple organisms call this program repeatedly for each one.\n\
A single table with the structure is created.\n\
\n\
parameters:\n\
--help        print this usage info\n\
--chunksize   this option is only relevant when loading a sequence file\n\
              the sequences in the input genome are split into chunks of this size so\n\
              that subsequent retrievals of small sequence ranges do not require to read\n\
              the complete - potentially much longer - chromosome. (<= 1000000, default " << chunksize << ")\n\
--noIdx       use this flag to suppress the building of indices on the database tables.\n\
              If you are going to load several genomes and/or hint files in a row, this option\n\
              is recommended to speed up the loading. But make sure to build indices with\n\
              --makeIdx after all genomes/hints are loaded. Otherwise, data retrieval operations\n\
              can be very slow.\n\
--makeIdx     use this flag to build the indices on the database tables after loading several\n\
              genomes and/or hint files with --noIdx. Only call this once for all species, e.g.\n\
              load2sqlitedb --makeIdx --dbaccess=database.db\n\
--clean       makes a clean load deleting existing hints/genome for the species from the database.\n\
              When called with a gff file, only the hints for the species are delete, but not the genome.\n\
              When called with a fasta file, both hints and genome for the species are deleted.\n\
\n\
examples:\n\
     load2sqlitedb --species=chicken --dbaccess=chicken.db chickengenome.fa\n\
     load2sqlitedb --species=chicken --dbaccess=chicken.db chickenhints.gff\n";
}
